pkg/icingadb/history/sync.go

```diff
@@ -32,274 +35,366 @@ func NewSync(db *icingadb.DB, redis *icingaredis.Client, logger *zap.SugaredLogg
-			for _, r2s := range redis2structs {
-				close(r2s)
-			}
-		}()
+// readFromRedis is the first stage of the history sync pipeline. It reads the history stream from Redis
+// and feeds the history entries into the next stage.
+func (s Sync) readFromRedis(ctx context.Context, key string, output chan<- redis.XMessage) error {
+	defer close(output)
 
-		xra := &redis.XReadArgs{
-			Streams: []string{stream, "0-0"},
-			Count:   bulkSize,
-			Block:   10 * time.Second,
-		}
+	xra := &redis.XReadArgs{
+		Streams: []string{"icinga:history:stream:" + key, "0-0"},
+		Count:   bulkSize,
+		Block:   10 * time.Second,
+	}
 
-		for {
-			cmd := s.redis.XRead(ctx, xra)
-			streams, err := cmd.Result()
+	for {
+		cmd := s.redis.XRead(ctx, xra)
+		streams, err := cmd.Result()
 
-			if err != nil && err != redis.Nil {
-				return icingaredis.WrapCmdErr(cmd)
-			}
+		if err != nil && err != redis.Nil {
+			return icingaredis.WrapCmdErr(cmd)
+		}
 
-			for _, stream := range streams {
-				for _, message := range stream.Messages {
-					xra.Streams[1] = message.ID
+		for _, stream := range streams {
+			for _, message := range stream.Messages {
+				xra.Streams[1] = message.ID
 
-					for _, r2s := range redis2structs {
-						select {
-						case <-ctx.Done():
-							return ctx.Err()
-						case r2s <- message:
-						}
-					}
+				select {
+				case output <- message:
+				case <-ctx.Done():
+					return ctx.Err()
 				}
 			}
 		}
 	}
 }
 
-// structifyStream structifies from redis2struct to struct2db.
-func structifyStream(
-	ctx context.Context, structifier structify.MapStructifier, redis2struct <-chan redis.XMessage,
-	struct2db chan<- contracts.Entity, rowIds map[contracts.Entity]string, rowIdsMtx *sync.Mutex,
-) func() error {
-	return func() error {
-		defer close(struct2db)
-
-		for {
-			select {
-			case <-ctx.Done():
-				return ctx.Err()
-			case message, ok := <-redis2struct:
-				if !ok {
-					return nil
+// deleteFromRedis is the last stage of the history sync pipeline. It receives history entries from the second to last
+// pipeline stage and then deletes the stream entry from Redis as all pipeline stages successfully processed the entry.
+func (s Sync) deleteFromRedis(ctx context.Context, key string, input <-chan redis.XMessage) error {
+	const logInterval = 20 * time.Second
+	const flushTimeout = 250 * time.Millisecond
+
+	var ids []string             // Used to collect a set of IDs before issuing a Redis XDEL command.
+	var count uint64             // Count of synced entries for periodic logging.
+	var timeout <-chan time.Time // Used to force a flush if ids after some time if the bulk size is not reached.
+
+	stream := "icinga:history:stream:" + key
+	logTicker := time.NewTicker(logInterval)
+	defer logTicker.Stop()
+
+	for {
+		timeoutReached := false
+		inputClosed := false
+
+		select {
+		case e, ok := <-input:
+			if ok {
+				ids = append(ids, e.ID)
+				if len(ids) == 1 {
+					timeout = time.After(flushTimeout)
```

Comment:
If you've not just C&Ped my code, but wrote this more or less from scratch, consider (re-)using an actual timer instead. Rationale: see time.After docs.